django/db/models/expressions.py

```diff
@@ -1711,14 +1711,24 @@ def as_sql(
             except EmptyResultSet:
                 continue
             except FullResultSet:
-                default_sql, default_params = compiler.compile(case.result)
+                default = case.result
                 break
             case_parts.append(case_sql)
             sql_params.extend(case_params)
         else:
-            default_sql, default_params = compiler.compile(self.default)
-        if not case_parts:
-            return default_sql, default_params
+            default = self.default
+        if case_parts:
+            default_sql, default_params = compiler.compile(default)
+        else:
+            if (
+                isinstance(default, Value)
+                and (output_field := default._output_field_or_none) is not None
+            ):
+                from django.db.models.functions import Cast
+
+                default = Cast(default, output_field)
+            return compiler.compile(default)
```

Comment:
```suggestion
            from django.db.models.functions import Cast

            output_field = default._output_field_or_none
            default = Cast(default, output_field)
            return compiler.compile(default)
```
It might be that I've not run enough tests or with the wrong database, but I don't think a test fails if I remove the if guard here. Perhaps we need another test case?

Reply:
I'm thinking through this, and I think we only want to apply the Cast to a Value. I'll try adding a test with a default that is an Expression.

Reply:
Is the default is always a Value if case_parts is empty?

Reply:
We could always cast but it would be unnecessary in most cases, it's only necessary when `isinstance(default, Value)` and particularly when `isinstance(default.value, int)` on Postgres.

Reply:
It's ok if It's not possible, just wanted to check 👍